### PR TITLE
shogun 4.0.0

### PR DESCRIPTION
**Pull request by saketkc**

### CODE REVIEW

**Comment by dpo** [comment]
I worked on a new `shogun` formula a while back but could never get all the interfaces to work. See #217.

**Comment by saketkc** [comment]
Thanks, will take a look!

**Comment by webdrone** [comment]
Hello,

I appreciate all your efforts of getting this to work -- I'd like to know if there's still a chance of getting shogun to install on brew or if I should just abandon all hope and use ports.

Could you give me a rough idea of what the issue is?

Best regards.

**Comment by dpo** [comment]
See https://github.com/Homebrew/homebrew-science/issues/217#issuecomment-49491413. If you don't need one of the interfaces that don't build, you can use my gist.

**Comment by saketkc** [comment]
There is. I have tonnes of deadlines this week, but I will try to finish it asap.

**Comment by webdrone** [comment]
@dpo oh thanks! I don't care about octave or R, I understand those are the ones that don't build, right? Despite that, your gist is for the 3.2 version, not 4.0, correct?
@saketkc Thank you, it would be great! but ofc prioritise your deadlines!

**Comment by dpo** [comment]
@webdrone Yes I haven't touched the gist in a while since the bug reports weren't getting any attention upstream.

**Comment by webdrone** [comment]
excellent, thank you! hope to see it working soon.

**Comment by saketkc** [comment]
This is good to go now.

File: shogun.rb [hunk 1]
```diff
@@ -12,72 +11,71 @@ class Shogun < Formula
...
-      # fix: "Checking for Mac OS vector library ... no"
-      ["src/configure", "src/shogun/mathematics/lapack.h"].each do |f|
-        inreplace f, "#include </System/Library/Frameworks/vecLib.framework",
-                     "#include <#{MacOS.sdk_path}/System/Library/Frameworks/vecLib.framework"
-      end
-    end
+  depends_on "homebrew/python/matplotlib" if build.with? "brewed-matplotlib"
+  depends_on "r" => :optional
+  depends_on "lua" => :optional
+  depends_on "octave" => :optional
```
Comment (dpo) [inline shogun.rb]: Shouldn't this be something like `depends_on "r" if build.with?("r-modular") || build.with?("r-static")`?

File: shogun.rb [hunk 2]
```diff
@@ -12,72 +11,71 @@ class Shogun < Formula
...
+  depends_on "readline" => :recommended
+  depends_on "nlopt" => :recommended
+  depends_on "eigen" => :recommended
+  depends_on "glpk" => :recommended
+  depends_on "lzo" => :recommended
+  depends_on "snappy" => :recommended
+  depends_on "xz" => :recommended  # provides lzma
+  depends_on "swig" => [:recommended, :build] # needed for dynamic python bindings
+  depends_on :python      => :recommended
+  depends_on "homebrew/python/numpy" if build.with? "brewed-numpy"
```
Comment (dpo) [inline shogun.rb]: Should there be a `with-brewed-numpy` option? In addition, `:python` should be built if the user selects either `numpy` or `matplotlib`.

Comment (saketkc) [inline shogun.rb]: Yes, haste made all this waste.

**Comment by saketkc** [comment]
Does this look good now?

**Comment by saketkc** [comment]
@dpo Would you want to test this out?

**Comment by dpo** [comment]
I will when I get a chance. We should exercise the interfaces.

**Comment by dpo** [comment]
So far, `--with-java-modular` and `--with-lua-modular` fail for me on Mavericks.

**Comment by saketkc** [comment]
Started buiding all options here: https://github.com/saketkc/brewer/tree/shogun

**Comment by dpo** [comment]
Great thanks! Here's the build-o-meter:
[Build Status](https://travis-ci.org/saketkc/brewer)

**Comment by dpo** [comment]
`--with-octave-modular`, `--with-r-modular` and `--with-python-modular` all fail for me too.

File: shogun.rb [hunk 3]
```diff
@@ -12,72 +11,72 @@ class Shogun < Formula
...
-      ["src/configure", "src/shogun/mathematics/lapack.h"].each do |f|
-        inreplace f, "#include </System/Library/Frameworks/vecLib.framework",
-                     "#include <#{MacOS.sdk_path}/System/Library/Frameworks/vecLib.framework"
-      end
-    end
+  depends_on "homebrew/python/matplotlib" if build.with? "brewed-matplotlib"
+  depends_on "r" if build.with?("r-static") || build.with?("r-modular")
+  depends_on "lua" if build.with?("lua-static") || build.with?("lua-modular")
+  depends_on "octave" if build.with?("octave-static") || build.with?("octave-modular")
+  depends_on "python" if build.with?("brewed-numpy") || build.with?("brewed-matplotlib")
```
Comment (dpo) [inline shogun.rb]: This should be `:python`

File: shogun.rb [hunk 4]
```diff
@@ -12,72 +11,72 @@ class Shogun < Formula
...
-  depends_on 'snappy' => :recommended
-  depends_on 'xz' => :recommended  # provides lzma
-  depends_on 'swig' => [:recommended, :build] # needef for dynamic python bindings
-  depends_on 'numpy' => :python # You may want `brew tap homebrew/python && brew install numpy`
-  depends_on 'matplotlib' => :python # You may want `brew tap homebrew/python && brew install matplotlib`
-  depends_on 'r' => :optional
-  depends_on 'lua' => :optional
-  depends_on 'octave' => :optional
+  option "with-csharp-modular", "Build with c# modular interface"
+  option "with-java-modular", "Build with java modular interface"
```
Comment (dpo) [inline shogun.rb]: The formula is also missing a `depends_on :java if ...` Have a look at https://gist.github.com/dpo/388b3d8aa2a030c055f3#file-shogun-rb. In that gist, I was specifying all the `cmake` arguments. It seemed ok, but there was (and still is) a problem with `colpack`.

**Comment by dpo** [comment]
The build is still broken if `colpack` is installed:
````
Undefined symbols for architecture x86_64:
  "ColPack::GraphColoring::GetVertexColors(std::__1::vector<int, std::__1::allocator<int> >&)", referenced from:
      shogun::CProbingSampler::precompute() in ProbingSampler.cpp.o
  "ColPack::GraphColoringInterface::Coloring(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)", referenced from:
      shogun::CProbingSampler::precompute() in ProbingSampler.cpp.o
````

**Comment by saketkc** [comment]
This is getting really tricky:
https://travis-ci.org/saketkc/brewer/builds/52505465

**Comment by dpo** [comment]
Ow:
```
Building CXX object gtest/CMakeFiles/gtest_main.dir/src/gtest_main.cc.o
{standard input}: Assembler messages:
{standard input}:4849: Error: no such instruction: `mo'
{standard input}: Error: open CFI at the end of file; missing .cfi_endproc directive
g++-4.8: internal compiler error: Killed (program cc1plus)
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-4.8/README.Bugs> for instructions.
make[2]: *** [src/shogun/CMakeFiles/libshogun.dir/statistics/NOCCO.cpp.o] Error 4
g++-4.8: internal compiler error: Killed (program cc1plus)
```

**Comment by saketkc** [comment]
Travis has a really strange behavior. I tried multiple gcc versions.It builds fine locally, at least a few of the options(I haven't tried all).

**Comment by webdrone** [comment]
many thanks for making this work! :+1:

**Pull request closed without merging**